OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala

```diff
@@ -0,0 +1,1413 @@
+                    if (!intentMethods.contains(call.name)) {
+                        return Some[(ObjectType, String)](call.declaringClass.asObjectType -> call.name)
+                    }
+                    return searchParamsStatic(call, tacCode)
+                }
+            }
+        }
+        Some[(ObjectType, String)](functionCall.declaringClass.asObjectType -> functionCall.name)
+    }
+
+    /**
+     * searches the parameters of functionCall to find the origin of the intent and return the class and method
+     * that returns the searched intent
+     *
+     */
+    def searchParamsVirtual(
+                             functionCall: MethodCall[DUVar[ValueInformation]],
+                             tacCode: AITACode[TACMethodParameter, ValueInformation]
+                           ): Option[(ObjectType, String)] = {
+        functionCall.params.foreach { p ⇒
+            if (p.asVar.value.toString.startsWith(intentStart.head) || p.asVar.value.toString.startsWith(intentStart(1))) {
+
+                val defIndex = p.asVar.definedBy.head
+                if (defIndex < 0) {
+                    return None
+                }
+                val defSiteStm = tacCode.stmts(defIndex)
+                if (defSiteStm.isAssignment && (defSiteStm.asAssignment.expr.isVirtualFunctionCall ||
+                  defSiteStm.asAssignment.expr.isStaticFunctionCall)) {
+
+                    val call = defSiteStm.asAssignment.expr.asFunctionCall
+                    if (!intentMethods.contains(call.name)) {
+                        return Some[(ObjectType, String)](call.declaringClass.asObjectType -> call.name)
+                    }
+                    return searchParamsStatic(call, tacCode)
+                }
+            }
+        }
+        Some[(ObjectType, String)](functionCall.declaringClass.asObjectType -> functionCall.name)
+    }
+
+    def getComponentType(
+                          name: String
+                        ): String = {
+        if (activityStartMethods.contains(name)) {
+            return activity
+        }
+        if (serviceMethods.contains(name)) {
+            return service
+        }
+        if (broadcastReceiverStartMethods.contains(name)) {
+            return receiver
+        }
+        undef
+    }
+
+    def reconstructIntent(
+                           c:           NonVirtualMethodCall[V],
+                           tacCode:     AITACode[TACMethodParameter, ValueInformation],
+                           useSites:    List[Int],
+                           m:           Method
+    ): Unit = {
+        val p = c.params
+
+        p.size match {
+            case 0 ⇒ //Intent()
+                val intent = new implicitIntent(m)
+                evaluateUseSites(tacCode, useSites, intent)
+            case 1 ⇒ //Intent(String action)
+                if (p.head.asVar.value.verificationTypeInfo == ObjectVariableInfo(ObjectType(stringString))) {
```

Comment:
There are constants available for ObjectTypes like java/lang/String: ObjectType.String. Please always use these as it is way more performant than looking them up by their name each time. For types that don't have constants in OPAL but are needed by you several times, consider looking up the ObjectType only once and store it as a constant in your analysis.